Publish artifacts and symbols one file at a time

epananth opened the pull request:
### To double check:

* [ ] The right tests are in and and the right validation has happened.  Guidance:  https://github.com/dotnet/core-eng/tree/master/Documentation/Validation

Issue -> https://github.com/dotnet/arcade/issues/6987

Both the old functionality to download the entire artifact folder is still present and can be turned on by enabling the artifact download task and switching UseStreamingPublishing= false in the publish-assets.yml 

There is definitely some redundant code, I'm hoping code review will help me make it better. 


NOTE : PdbArtifacts are still downloaded like we did earlier, cos there is not way to get the file names before hand to download them from PdbArtifacts directory.

epananth commented:
Some tests I ran 

Runtime Latest build -> https://dnceng.visualstudio.com/internal/_build/results?buildId=1040969&view=results
vs Runtime build with new hosted pool -> https://dnceng.visualstudio.com/internal/_build/results?buildId=1040959&view=results

Other promotion builds 
dotnet-sdk - 
https://dnceng.visualstudio.com/internal/_build/results?buildId=1040973&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a

dotnet-winforms - 
https://dnceng.visualstudio.com/internal/_build/results?buildId=1040984&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a

epananth commented:
More test results from day before 

 https://dnceng.visualstudio.com/internal/_build/results?buildId=1039004&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a - 20 m 48 seconds
                 https://dnceng.visualstudio.com/internal/_build/results?buildId=1039011&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a - 20 m 46 seconds
                 https://dnceng.visualstudio.com/internal/_build/results?buildId=1039034&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a - 21 m 42 seconds

this download with the new hosted pool took 31 minutes https://dnceng.visualstudio.com/internal/_build/results?buildId=1038957&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=cecdd0ae-1666-570d-4521-d44005725bca, compared to the whole thing taking https://dev.azure.com/dnceng/internal/_build/results?buildId=1039004&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a 20 m 48 seconds..

epananth commented:
Odd one 

Runtime took 31 mins 
https://dnceng.visualstudio.com/internal/_build/results?buildId=1040978&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a

epananth commented:
I ran tests for different promotions parallelly, and they all succeeded. I have also added retry logic for downloading files. So I think I covered most of the tests, we had discussed.

riarenas commented on eng/publishing/v3/publish-assets.yml:
```diff
@@ -87,7 +106,12 @@ jobs:
           /p:SymWebToken=$(symweb-symbol-server-pat)
           /p:BuildQuality='${{ parameters.buildQuality }}'
-
+          /p:AzdoApiToken='$(dn-bot-dotnet-build-rw-code-rw)'
```
[nit] this secret name doesn't seem to follow the general guidelines for what a PAT should be called. The dotnet portion should tell me which azure devops orgs this token will work for: https://github.com/dotnet/core-eng/blob/main/Documentation/Secret-Management/secret-naming-conventions.md#visual-studio-team-system-personal-access-token-pat

epananth replied:
created dn-bot-all-orgs-build-rw-code-rw

riarenas reviewed (commented):
First round of feedback. Still need to finish looking at all the files.

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +659,121 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+                    string containerId = "";
+                    foreach (var bd in buildArtifacts.value)
+                    {
+                        if (string.Equals(bd.name, "BlobArtifacts"))
```
Oh, it looks like GetContainerID is actually GetBlobArtifactsContainerId?

epananth replied:
It used to download all the artifacts so I used a generic name. ContainerId is the same for packages and blobs

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +659,121 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+                    {
+                        if (string.Equals(bd.name, "BlobArtifacts"))
+                        {
+                            string[] segment = bd.resource.data.Split('/');
```
Why the split? what does this data actually look like?

epananth replied:
this is how it looks like assets/symbols/Microsoft.Arcade.Common.6.0.0-beta.21119.10.symbols.nupkg, so I just get the name of the file

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +659,121 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        public async Task<bool> DownloadFileAsync(HttpClient client, string artifact, string containerId,
+            string fileName, string path)
+        {
+            int attempt = 0;
```
This should use the ExponentialRetryHandler available in https://github.com/dotnet/arcade/blob/main/src/Common/Microsoft.Arcade.Common/ExponentialRetry.cs

epananth replied:
updated it

epananth replied:
updated it

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+            }
+
+            string containerId = GetContainerId().Result;
+
```
extra blank line

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                }
+            }
+
+            string containerId = GetContainerId().Result;
```
Same comment about using await

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+            string containerId = GetContainerId().Result;
+
+
+            using HttpClientHandler handler = new HttpClientHandler()
```
Same comment about just making this the general HttpClientHandler for publishing instead of passing it in to the CreateHttpClient function

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+
+                        string id;
+                        string version;
+                        using (var packageReader = new NuGet.Packaging.PackageArchiveReader(localBlobPath))
```
Why the full namespace here? is there a conflict with PackageArchiveReader?

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+            }
+
+        }
+
```
extra line

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+            string feedName = parsedUri.Groups["feed"].Value;
+            using (var clientThrottle = new SemaphoreSlim(MaxClients, MaxClients))
+            {
+                using (HttpClient httpClient = new HttpClient(new HttpClientHandler
```
Why is this HttpClient created differently than the others?

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                    {
+                        clientThrottle.Release();
+                    }
+
```
extra line

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                    try
+                    {
+                        // Wait to avoid starting too many processes.
+                        await clientThrottle.WaitAsync();
+                        await PushNugetPackageAsync(feedConfig, httpClient, localBlobPath, id,
+                            version, feedAccount, feedVisibility, feedName);
```
I don't quite understand this part. It looks like you're awaiting the PushNugetPackageAsync call so only one is happening at a time, so what exactly would start too many processes that you need the semaphore?

epananth replied:
redid this

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+
+                }
+            }
+        }
```
missing blank line between methods

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                PassIfExistingItemIdentical = true
+            };
+
+
```
extra blank line

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+            };
+
+
+            using HttpClientHandler handler = new HttpClientHandler()
```
same comment about this as before.

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                        return;
+                    }
+
+                    TryAddAssetLocation(package.Id, package.Version, buildAssets, feedConfig,
```
Do we not care if the location was not added?

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -849,6 +1171,163 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+
+                    TryAddAssetLocation(package.Id, package.Version, buildAssets, feedConfig,
+                        AddAssetLocationToAssetAssetLocationType.NugetFeed);
+                    IEnumerable<string> item = new List<string>();
+                    item.ToList().Add(localPackagePath);
+                    await blobFeedAction.PushToFeedAsync(item, pushOptions);
```
This seems odd to me, why are we generating a list with one whole item? Shouldn't we create an override that takes a single item?

epananth replied:
https://github.com/dotnet/arcade/blob/c5a82c9ee874ea5dd2831b5f2b34e9ccc6a50a56/src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs#L346

PushCommand.RunAsync actually takes a list, so I used a list. I don't think we need PushToFeedAsync anymore, so I removed that method.

riarenas reviewed (commented):
Second pass. Two files to go

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BuildArtifacts.cs:
```diff
@@ -0,0 +1,35 @@
+using System.Collections.Generic;
```
done

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BuildArtifacts.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Microsoft.DotNet.Build.Tasks.Feed
+{
+    public class Properties
```
its part of the API return, so I kept it as is

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -315,6 +341,135 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+                var blobs = blobsPerCategory.Value;
+                foreach (var blob in blobs)
+                {
+                    var segments = blob.ToString().Split('/');
```
changed it

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +659,121 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
             };
         }
 
+        public HttpClient CreateHttpClient(HttpClientHandler handler, string accountName, string projectName = null, string versionOverride = null, string baseAddressSubpath = null)
```
renamed it as createAzdo client

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +659,121 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+                    string uri =
+                        $"{AzureDevOpsBaseUrl}/{AzureDevOpsOrg}/{AzureProject}/_apis/build/builds/{BuildId}/artifacts?api-version={AzureDevOpsFeedsApiVersion}";
+                    using HttpRequestMessage getMessage = new HttpRequestMessage(HttpMethod.Get, uri);
+                    using HttpResponseMessage response = await client.SendAsync(getMessage);
```
changed it

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -315,6 +341,135 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+            IEnumerable<string> filesToSymbolServer = null;
+            using HttpClientHandler handler = new HttpClientHandler()
+                {
+                    AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
```
moved the handler inside CreateAzdoClient method

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -315,6 +341,135 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+
+            var symbolCategory = TargetFeedContentType.Symbols;
+
+            string containerId = GetContainerId().Result;
```
updated it

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -442,17 +597,42 @@ protected async Task HandlePackagePublishingAsync(Dictionary<string, HashSet<Ass
                         }
 
-                        switch (feedConfig.Type)
+                        if (!UseApiOverride)
```
redid this. Hopefully this is a better version

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +659,121 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+                        $"{AzureDevOpsBaseUrl}/{AzureDevOpsOrg}/{AzureProject}/_apis/build/builds/{BuildId}/artifacts?api-version={AzureDevOpsFeedsApiVersion}";
+                    using HttpRequestMessage getMessage = new HttpRequestMessage(HttpMethod.Get, uri);
+                    using HttpResponseMessage response = await client.SendAsync(getMessage);
+                    response.EnsureSuccessStatusCode();
```
added some exception handling..

mmitche replied:
In the exceptional case, it looks like it will still return an empty string. Will this cause another unhandled exception later on? Maybe we should add some tests for this. Logging the error is fine, but we should either gracefully continue on, or fail fast.

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +659,121 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+                    using HttpRequestMessage getMessage = new HttpRequestMessage(HttpMethod.Get, uri);
+                    using HttpResponseMessage response = await client.SendAsync(getMessage, timeoutTokenSource.Token);
+                    if (response.StatusCode == HttpStatusCode.NotFound ||
+                        response.ReasonPhrase.StartsWith("The requested URI does not represent any resource on the server.", StringComparison.OrdinalIgnoreCase))
```
changed it to use exponential retry..

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +659,121 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+                    using (var fs = new FileStream(path, FileMode.Create,
+                        FileAccess.ReadWrite, FileShare.ReadWrite))
+                    {
+                        using (var stream = await response.Content.ReadAsStreamAsync())
```
thanks for pointing this out. Updated it

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -377,15 +463,95 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+                        symbolLog.AppendLine($"Publishing symbol packages to {serverPath}:");
+                        symbolLog.AppendLine(
+                            $"Performing symbol publishing...\nSymbolServerPath : ${serverPath} \nExpirationInDays : {ExpirationInDays} \nConvertPortablePdbsToWindowsPdb : false \ndryRun: false \nTotal number of symbol files : {fileEntries.Length} ");
+                        await PublishSymbolsHelper.PublishAsync(
```
Does this throw or error in any way? Is it possible we could say that symbols were published but that be false?

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -619,20 +909,54 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
-                    string localPackagePath = Path.Combine(PackageAssetsBasePath, $"{package.Id}.{package.Version}.nupkg");
+                    var packageFilename = $"{package.Id}.{package.Version}.nupkg";
+                    string temporaryPackageDirectory =
+                        Path.GetFullPath(Path.Combine(StagingDir, @"..\", Guid.NewGuid().ToString()));
```
Same comment about the StagingDir + .. and the safety of that

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -821,110 +1146,270 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                foreach (var package in packagesToPublish)
+                {
+                    string temporaryPackageDirectory =
+                        Path.GetFullPath(Path.Combine(StagingDir, @"..\", Guid.NewGuid().ToString()));
```
StagingDir thingy

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -821,110 +1146,270 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                foreach (var blob in blobsToPublish)
+                {
+                    string temporaryBlobDirectory =
+                        Path.GetFullPath(Path.Combine(StagingDir, @"..\", Guid.NewGuid().ToString()));
```
StagingDir thingy

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -821,110 +1146,270 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+
+                foreach (var blob in blobsToPublish)
+                {
+                    string temporaryBlobDirectory =
```
We do this whole, get the path to the temp directory, ensure it exists thing a lot. Could this be pulled out into a function so that it's not duplicated all over the place?

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs:
```diff
@@ -183,6 +183,17 @@ public string BuildQuality
         }
+        public string AzdoApiToken {get; set;}
+
+        public string StagingDir { get; set;}
```
Can you help me understand what will happen if these things aren't supplied? Can you add a comment on UseApiOverride to let the user know if true = use the old behavior or false = use the old behavior?

mmitche replied:
Yep, commenting the new task parameters in general would be great

adiaaida reviewed (commented):
Initial comments

mmitche commented on eng/publishing/v3/publish-assets.yml:
```diff
@@ -87,7 +106,12 @@ jobs:
+          /p:AzdoApiToken='$(dn-bot-dotnet-build-rw-code-rw)'
+          /p:StagingDir='$(Build.ArtifactStagingDirectory)/'
+          /p:BuildId='$(AzDOBuildId)'
+          /p:AzureDevOpsOrg='dnceng'
```
Should this be harcoded?

epananth replied:
it was test stuff I forgot to change, I get this info from AzDOAccount. Changed this

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs:
```diff
@@ -140,6 +138,66 @@ public async Task PublishToFlatContainerAsync(IEnumerable<ITaskItem> taskItems,
             }
         }
 
+        public async Task UploadAssetAsync(
```
Is the point of this method to upload without the throttle lock? If that's the only difference, I'd suggest accepting null for the client throttle, and if null, avoiding the wait/release.

epananth replied:
there are 2, 
1) Uses throttle for the list and other override method does not 
2) uses a list of blob paths and other override method uses only a blob path 

didn't have a better way to refactor this.

mmitche replied:
Where is the list of blob paths in the override? I'm missing it.

epananth replied:
omg! yes.. I missed this, I was talking about another method.. I will change this like you initially suggested!!

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -317,57 +345,115 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+        /// <param name="msdlToken"></param>
+        /// <param name="symWebToken"></param>
+        /// <param name="symbolPublishingExclusionsFile"></param>
+        /// <param name="publishSpecialClrFiles"></param>
```
Can you add comments for these parameters?

riarenas replied:
This comment was resolved, but the parameters are still not commented. Reopened

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -377,15 +463,95 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+            Dictionary<string, HashSet<Asset>> buildAssets,
+            string temporarySymbolsLocation = null)
+        {
+            if (!UseApiOverride)
```
I suggest we pick a better name for `UseApiOverride` that more accurately reflects its purpose. Perhaps `UseStreamingPublishing`

adiaaida replied:
+1 on this. I'm still not sure what UseApiOverride is supposed to do.

epananth replied:
Sure. Changing this,

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -377,15 +463,95 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+        {
+            if (!UseApiOverride)
+            {
+                StringBuilder symbolLog = new StringBuilder();
```
I'd factor this code out into a separate method so that this method is just a shell with a choice on publishing methodology and two calls.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +647,122 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
             };
         }
 
+        public HttpClient CreateAzdoClient(string accountName,  bool download, string projectName = null, string versionOverride = null, string baseAddressSubpath = null)
```
Can you add a method comment header to this indicating what the purpose of this client is (which will explain the various properties below)

mmitche replied:
Is `baseAddressSubpath` used anywhere? If not, can it be removed?

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +647,122 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+
+                    foreach (var artifact in buildArtifacts.value)
+                    {
+                        if (string.Equals(artifact.name, "BlobArtifacts"))
```
Does this work if PackageArtifacts is what is getting streamed in?

epananth replied:
yes, containerId is the same for packageArtifacts and blobArtifacts.

mmitche replied:
Can you add a comment here detailing that since that fact is not obvious? Also, I think there is a constant for "BlobArtifacts" already (and if not, we should introduce one)

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +647,122 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        /// <summary>
+        /// Download artifact file using Azure API
+        /// </summary>
+        /// <param name="client"></param>
```
Can you add info on the parameters here?

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +647,122 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+            }).ConfigureAwait(false);
+            if (!success)
+            {
+                throw new Exception($"Failed to download local file '{path}' after {RetryHandler.MaxAttempts} attempts.  See inner exception for details.");
```
Is this supposed to pass `mostRecentlyCaughtException` as the inner exception?

mmitche replied:
Are there tests for when this throws in the various artifact cases (package publishing, symbol publishing, and blob publishing?)

epananth replied:
Nope nothing so far, I added this kind of retry after Azdo error happened a week or 2 ago..

epananth replied:
Talking about the unit test, It is going to be a part of a separate PR.

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -317,57 +345,115 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+                if (name.Contains(".symbols.nupkg"))
+                {
+                    var segments = name.Split('/');
+                    var fileName = segments[segments.Length - 1];
```
this is the path /assets/blobs/file.symbols.nupkg, I tried using Path.GetFileName() and it failed so I added this.

adiaaida replied:
Wait, this is like the path is D:\blah\blah\blah/assets/blobs/file.symbols.nupkg? This could get REALLY dicey if these paths are ever fixed to use consistent slashes.

epananth replied:
Nope, this is coming from the merged manifest. Where we have something like this 
```
  <Blob Id="assets/symbols/Microsoft.Arcade.Common.6.0.0-beta.21119.10.symbols.nupkg" NonShipping="true" />

adiaaida replied:
Ok. I'm going to do some digging. I feel like Path.GetFileName should do the right thing, assuming that this is getting "Microsoft.Arcade.Common.6.0.0-beta.21119.10.symbols.nupkg".

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -614,27 +960,77 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+            string containerId = await GetContainerIdAsync();
+            string localPackagePath = "";
+            using HttpClient client = CreateAzdoClient(AzureDevOpsOrg, true);
+            foreach (var package in packagesToPublish)
```
I don't think it needs to be altered in this PR, but there may be some performance left on the table here (the push isn't happening in parallel). It may not make as much of a difference compared to the old way, because of the lack of a mass-download step prior to upload, but something to think about.

epananth replied:
https://github.com/dotnet/arcade/issues/7180 created and issue to track this

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -880,15 +1360,160 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+
+            foreach (var blob in blobsToPublish)
+            {
+                string temporaryBlobDirectory = CreateTemporaryDirectory();
```
Does this temp dir need to be per-blob, or per set of blobs?

epananth replied:
I tried with a general directory like tempBlob/newBlob , but when I tried to delete the file after upload I got a lot of "used by another process" IO exceptions. So I had to create a directory for every file, but I delete it as soon as I upload it.

mmitche replied:
Hmmm. Is it possible that the file was being used in multiple places? Seems odd that separate directories + file deletes had different behavior than single directory + file deletes.

epananth replied:
After numerous tries, I just gave up and created one directory per file. I could not figure out which process was causing that error. :( 
I thought the upload did not finish and the delete was trying to delete the file before that. But that was not the case. It was something else. 
Is this current approach acceptable? I do delete the files and folders right after the upload, but this will cause a problem if the files and folders do not get deleted. :(

mmitche replied:
Yeah, current approach is fine, just odd that it works over a single dir.

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -974,6 +1591,101 @@ private BlobFeedAction CreateBlobFeedAction(TargetFeedConfig feedConfig)
+        {
+            if (Directory.Exists(temporaryLocation))
+            {
+                Directory.Delete(temporaryLocation);
```
In cases where the files inside couldn't be deleted, I think this may also throw. Probably should catch and log a warning.

adiaaida commented on eng/publishing/v3/publish-assets.yml:
```diff
@@ -22,12 +22,15 @@ jobs:
     - task: DownloadBuildArtifacts@0
       displayName: Download Build Assets
       continueOnError: true
+      enabled: false
```
Is this meant to be false?

epananth replied:
yes, we do not download the packageArtifact, blobArtifacts folder anymore, so this is set to be false.

adiaaida replied:
So why not delete the entire download step?

epananth replied:
So incase something fails in future, we can turn back this on. Since we support both the functionalities its good to have this, than remove it, I feel.

riarenas replied:
re-enabling it still requires a PR, and might also require reverting some other stuff, so I think we might as well delete it and bring it back if we need it again? I don't think we're really saving any time to fix a failure by keeping it in a disabled state

epananth replied:
I was thinking about this, if we remove this then we will have to make PackageBaseBath, BlobBasePath optional in older publishing infra.. I think its not good to do that.. I think we should keep both..

riarenas replied:
So you're saying you're going to enable it again?

Having a disabled task is basically the same thing as commenting out production code, there's no reason to do it because source control exists and getting anything back is easy.

epananth replied:
I removed the task, I was worried if I remove the task the null checks for older infra might fail. But it did not so, I am going to remove it.

adiaaida commented on eng/publishing/v3/publish-assets.yml:
```diff
@@ -87,7 +106,12 @@ jobs:
           /p:BuildQuality='${{ parameters.buildQuality }}'
-
+          /p:AzdoApiToken='$(dn-bot-all-orgs-build-rw-code-rw)'
+          /p:ArtifactsBasePath='$(Build.ArtifactStagingDirectory)/'
```
is the extra slash at the end necessary?

epananth replied:
Actually I copied the paths we already had in the yaml. But I checked without the with or without the / it will work.

adiaaida commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/PublishArtifactsInManifest.proj:
```diff
@@ -147,7 +151,14 @@
-      SymbolsFeedOverride="$(SymbolsFeedOverride)" />
+      SymbolsFeedOverride="$(SymbolsFeedOverride)" 
+      AzdoApiToken="$(AzdoApiToken)"
+      ArtifactsBasePath="$(ArtifactsBasePath)"
```
do we need to do anything if no ArtifactsBasePath is specified? Set a default? Or is empty ok?

epananth replied:
Artifact base path has mergedManifest in it. So its never empty.

adiaaida replied:
Not the directory, but the string itself. That's not guaranteed. Also, the publishing out of arcade isn't the only place where this proj file is being used (We are also using it in Stage-DotNet for later publishing). We need to make sure that these changes will not affect other places that use it/that the proj file is still backwards compatible. If ArtifactsBasePath is ok being empty (the string, not the directory), that's fine. But if it isn't, then that needs to be codified in this file.

riarenas replied:
We should probably just add a default if it's not set.

epananth replied:
adding default for ArtifactBasePath

riarenas replied:
I think the error in LN91 will still happen before you assign the default?

epananth replied:
I removed that error line and added default path equal to BlobBasepath, sorry for the confusion I had this in my test branch before

adiaaida commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/PublishArtifactsInManifest.proj:
```diff
@@ -147,7 +151,14 @@
+      BuildId="$(BuildId)"
+      AzureDevOpsOrg="$(AzureDevOpsOrg)"
+      AzureProject="$(AzureProject)"
+      UseStreamingPublishing="$(UseStreamingPublishing)"/>
```
We should probably have a conditional above (ie line 109) to set this to false if it isn't passed in

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs:
```diff
@@ -140,6 +138,66 @@ public async Task PublishToFlatContainerAsync(IEnumerable<ITaskItem> taskItems,
+                relativeBlobPath = $"{recursiveDir}{fileName}";
+            }
+
+            relativeBlobPath = $"{RelativePath}{relativeBlobPath}".Replace("\\", "/");
```
WE should use Path.Combine

adiaaida replied:
Er... maybe not. it's a relative path. Lemme think on this.

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs:
```diff
@@ -140,6 +138,66 @@ public async Task PublishToFlatContainerAsync(IEnumerable<ITaskItem> taskItems,
+            {
+                string fileName = Path.GetFileName(item.ItemSpec);
+                string recursiveDir = item.GetMetadata("RecursiveDir");
+                relativeBlobPath = $"{recursiveDir}{fileName}";
```
Is this guaranteed to have the correct slashes in it? I worry that if a slash is missing from RecursiveDir, we could get in a sticky spot.

epananth replied:
yes, it has always been like this. So far we have not faced any problem with it.

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs:
```diff
@@ -140,6 +138,66 @@ public async Task PublishToFlatContainerAsync(IEnumerable<ITaskItem> taskItems,
+                relativeBlobPath = $"{recursiveDir}{fileName}";
+            }
+
+            relativeBlobPath = $"{RelativePath}{relativeBlobPath}".Replace("\\", "/");
```
We should also probably confirm that RelativePath ends with a trailing slash.

epananth replied:
I reverted it and changed it back to the way it was earlier. We check if the blobPath exists or not in publishArtifactsInManifestV3

epananth replied:
also added a check

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs:
```diff
@@ -140,6 +138,66 @@ public async Task PublishToFlatContainerAsync(IEnumerable<ITaskItem> taskItems,
+                throw;
+            }
+        }
         public async Task UploadAssetAsync(
```
There is a lot of code that looks duplicated between this UploadAssetAsync function and the above UploadAssetAsync function. We should look at factoring out shared code so that if we need to update it, we don't have to update in 2 places.

epananth replied:
Yes, I don't think I will need this anymore.. I am going revert this change.

epananth replied:
refactoring this

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs:
```diff
@@ -183,6 +183,21 @@ public string BuildQuality
+
+        /// <summary>
+        /// If true, uses Azdo Api to download artifacts and symbols files one file at a time during publishing process.
+        /// If it is set to false, then artifacts and symbols are downloaded in PackageArtifacts and BlobArtifacts directory before publishing. 
```
Does this mean that PublishArtifactsInManifest will do the downloading itself? That's, I think, going to severely limit my ability to use this task in the Stage-DotNet pipeline, as we are republishing newly signed, etc files that were generated by the pipeline. Honestly, I think if false, this code should assume that the files should have already been downloaded and exist in the expected places (ie, the prior behavior).

epananth replied:
You should be able to use the older way of publishing by enabling downloadArtifacts and setting UsingStreamingPublishing to true. That way you will not use newer way of publishing.

cc: @mmitche

adiaaida replied:
Ok great. That's what I wanted to make sure of.

For the record, we don't actually use any of this yaml. We just call PublishArtifactsInManifest.proj. What I'm trying to make sure that when we take this arcade update, we won't need to make any changes to our scripts where we publish in addition to taking the update.

adiaaida replied:
Oh.... I didn't see the last half of the comment. *sigh* sorry.

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestV3.cs:
```diff
@@ -199,16 +189,31 @@ public override async Task<bool> ExecuteAsync()
+                    EnsureTemporaryDirectoryExists(temporarySymbolsLocation);
+                    DeleteTemporaryFiles(temporarySymbolsLocation);
+
+                    //Copying symbol files to temporary location is required because the symUploader API needs read/write access to the files,
```
```suggestion
                    // Copying symbol files to temporary location is required because the symUploader API needs read/write access to the files,
```

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestV3.cs:
```diff
@@ -199,16 +189,31 @@ public override async Task<bool> ExecuteAsync()
+                    DeleteTemporaryFiles(temporarySymbolsLocation);
+
+                    //Copying symbol files to temporary location is required because the symUploader API needs read/write access to the files,
+                    //since we publish blobs and symbols in parallel this will cause IO exceptions.
```
```suggestion
                    // since we publish blobs and symbols in parallel this will cause IO exceptions.
```

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestV3.cs:
```diff
@@ -199,16 +189,31 @@ public override async Task<bool> ExecuteAsync()
                     HandleBlobPublishingAsync(buildAssets),
                     HandleSymbolPublishingAsync(PdbArtifactsBasePath, MsdlToken,
-                        SymWebToken, SymbolPublishingExclusionsFile, temporarySymbolsLocation, PublishSpecialClrFiles)
+                        SymWebToken, SymbolPublishingExclusionsFile, PublishSpecialClrFiles, buildAssets ,temporarySymbolsLocation)
```
```suggestion
                        SymWebToken, SymbolPublishingExclusionsFile, PublishSpecialClrFiles, buildAssets, temporarySymbolsLocation)
```

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestV3.cs:
```diff
@@ -199,16 +189,31 @@ public override async Task<bool> ExecuteAsync()
                     return false;
                 }
 
+                string temporarySymbolsLocation ="";
```
```suggestion
                string temporarySymbolsLocation = "";
```

epananth commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/PublishArtifactsInManifest.proj:
```diff
@@ -138,7 +138,14 @@
-      SymbolsFeedOverride="$(SymbolsFeedOverride)" />
+      SymbolsFeedOverride="$(SymbolsFeedOverride)" 
+      AzdoApiToken="$(AzdoApiToken)"
+      StagingDir="$(StagingDir)"
```
added that..

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs:
```diff
@@ -140,6 +138,66 @@ public async Task PublishToFlatContainerAsync(IEnumerable<ITaskItem> taskItems,
+            if (relativeBlobPath.Contains("//"))
+            {
+                Log.LogError(
+                    $"Item '{item.ItemSpec}' RelativeBlobPath contains virtual directory " +
```
I think it should be starts with //, I will change this..

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs:
```diff
@@ -332,7 +342,7 @@ private ISleetFileSystem GetAzureFileSystem(LocalCache fileCache)
         }
 
-        private async Task<bool> PushAsync(IEnumerable<string> items, PushOptions options)
+        public async Task<bool> PushAsync(IEnumerable<string> items, PushOptions options)
```
It looks like this method is basically wrapped by https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs#L103, Is there a reason to make this public instead of using that?

epananth replied:
I am calling PushAsync in PublishArtifactsInManifestBase, before it was called only in BlobFeedAction

riarenas replied:
Yes, what I'm saying is that there's another method (the one I link in my comment) that is already public and calls into this private method that you're making public.

My question is whether that is intentional, or just that you didn't see the other one.

epananth replied:
Yes other method which calls PushAsync which is PushToFeedAsync, does a sanity check to see if there are any duplicated in the ItemsList. Since I'm only pushing one item at a time. That part is not required. So I called PushAsync directly

riarenas replied:
So, looking at https://github.com/dotnet/arcade/blob/4535393d49a652c05fe7f7b2a4a752fca9ee1f45/src/Microsoft.DotNet.Build.Tasks.Feed/src/BlobFeedAction.cs#L103-L125 it seems like it just checks that the items to push is not empty and includes some error handling, which I don't think is a bad thing. I think I'd prefer if we used it over changing the visibility of this one that doesn't have any parameter validation. When it was private that was mostly fine because the class controlled all callers, but by making it public you're now opening it so that anyone can call it.

Alternatively, you could add parameter validation and error handling, but then you'd end up with pretty much the same method as the one I'm saying to use.

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -317,7 +353,142 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+        /// <param name="symbolPublishingExclusionsFile"></param>
+        /// <param name="publishSpecialClrFiles"></param>
+        /// <param name="buildAssets"></param>
+        /// <returns></returns>
```
[nit] the empty tags are not particularly useful
```suggestion
```

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -317,7 +353,142 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+                    localSymbolPath = Path.Combine(temporarySymbolsDirectory, symbol);
+                    await DownloadFileAsync(client, BlobArtifacts, containerId, symbol, localSymbolPath);
+
+                    Log.LogMessage(MessageImportance.Low, $"Local Symbol path to downloaded file {localSymbolPath}");
```
Should this be appending to the StringBuilder instead?

epananth replied:
using StringBuilder instead

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -315,6 +341,135 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+            }
+
+            // publishing pdb artifacts 
+            if (Directory.Exists(pdbArtifactsBasePath))
```
I think my question still stands? Is it OK to not say anything if the directory doesn't exist? (The answer can totally be yes if that's the case)

mmitche replied:
Yeah, PDBArtifacts are not all that common. I believe the typical case is not having any.

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +694,136 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        /// </summary>
+        /// <param name="accountName">Name of the org</param>
+        /// <param name="download">If client is used for downloading the artifact then AutomaticDecompression has to be set, else it is not required</param>
+        /// <param name="projectName">If it is internal or public</param>
```
[nit] This is a string, so it can be anything and there is nothing restricting the values to those two. It's pretty much just the azure devops project name

adiaaida replied:
If you want it to be restricted, you could make it an enum.

epananth replied:
Updated the description

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +694,136 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        /// <param name="accountName">Name of the org</param>
+        /// <param name="download">If client is used for downloading the artifact then AutomaticDecompression has to be set, else it is not required</param>
+        /// <param name="projectName">If it is internal or public</param>
+        /// <param name="versionOverride">For downloading file we use Version 4.1-preview.4 for other requests we use 6.0</param>
```
[nit] This remark doesn't really belong in the parameter documentation. You just expect a version override, and it's up to the caller to pass the right one for the API they want. Perhaps what we should document is the default value we use if we don't get a versionOverride.

epananth replied:
Updated param description

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +694,136 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        /// Gets the container Id, that is going to be used in another API call to download the assets
+        /// ContainerId is the same for PackageArtifacts and BlobArtifacts
+        /// </summary>
+        /// <returns></returns>
```
This method actually returns something, so it should say what it returns

epananth replied:
returns containerId added it

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +694,136 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        /// <param name="download">If client is used for downloading the artifact then AutomaticDecompression has to be set, else it is not required</param>
+        /// <param name="projectName">If it is internal or public</param>
+        /// <param name="versionOverride">For downloading file we use Version 4.1-preview.4 for other requests we use 6.0</param>
+        /// <returns></returns>
```
[nit] this returns something

epananth replied:
added return

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +694,136 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        /// ContainerId is the same for PackageArtifacts and BlobArtifacts
+        /// </summary>
+        /// <returns></returns>
+        private async Task<string> GetContainerIdAsync()
```
[nit] This seems to be totally specific to getting the BlobArtifacts containerID, maybe the name of the helper should reflect that?

mmitche replied:
Based on Epsi's tests, It seems like the container Id for all artifacts will be identical. BlobArtifacts just happens to be what gets looked up.

@epananth It might be safer to pass in the artifact name as a parameter and not assume they will always be the same. I wonder whether retries could cause some kind of split (e.g. if PackageArtifacts is only uploaded the second time through)

riarenas replied:
>Based on Epsi's tests, It seems like the container Id for all artifacts will be identical. BlobArtifacts just happens to be what gets looked up.

Interesting. Didn't know that. I think I also really like the idea of just passing in the artifact name as a parameter.

epananth replied:
I added the Artifact name a param to be passed. So there should be no confusion on this now.

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +694,136 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        /// Download artifact file using Azure API
+        /// </summary>
+        /// <param name="client">Azdo client</param>
+        /// <param name="artifact">If it is PackageArtifacts or BlobArtifacts</param>
```
There's nothing that restricts this value from being different from the two you seem to expect, should we have an enum, or fail the method quickly if the values are not correct?

epananth replied:
Made this as an enum

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -439,19 +650,23 @@ protected async Task HandlePackagePublishingAsync(Dictionary<string, HashSet<Ass
                         {
                             case FeedType.AzDoNugetFeed:
-                                publishTasks.Add(PublishPackagesToAzDoNugetFeedAsync(filteredPackages, buildAssets, feedConfig));
+                                publishTasks.Add(PublishPackagesToAzDoNugetFeedAsync(filteredPackages, buildAssets,
```
This applies to everywhere where you're chopping at the middle of the line

epananth replied:
changed this

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -505,17 +850,24 @@ protected async Task HandleBlobPublishingAsync(Dictionary<string, HashSet<Asset>
                         }
 
-                        switch (feedConfig.Type)
+                        if (!UseStreamingPublishing)
```
No else? just making sure we're not missing something here if this is set

epananth replied:
I did not remove the old code where had multiple switch statements. So I removed it now.  The use of streaming or not condition is inside the switch statement now.

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -838,20 +1257,86 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                }
+            }
+        }
+        private async Task PushBlobToNugetFeed(TargetFeedConfig feedConfig, string localBlobPath, string id,
```
[nit] missing blank line between methods

epananth replied:
added new line

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -614,27 +965,77 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                }
+                TryAddAssetLocation(package.Id, package.Version, buildAssets, feedConfig,
+                    AddAssetLocationToAssetAssetLocationType.NugetFeed);
+                await PushPackageToNugetFeed(feedConfig, localPackagePath, package.Id, package.Version);
```
Each of these calls is going to instantiate a new HttpClient, which is an anti-pattern. Could this method itself instantiate it and pass it around instead?

epananth replied:
Added a client before the method and passing that around instead of creating a new client every time

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -317,7 +353,142 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+            var symbolCategory = TargetFeedContentType.Symbols;
+            string containerId = await GetContainerIdAsync();
+            HashSet<string> symbolsToPublish = new HashSet<string>();
+            //Get all the symbol file names
```
nit: new line above here.

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -317,7 +353,142 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+                    symbolsToPublish.Add(Path.GetFileName(name));
+                }
+            }
+            HashSet<TargetFeedConfig> feedConfigsForSymbols = FeedConfigs[symbolCategory];
```
Nit: new line above this line

epananth replied:
added new line

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -317,7 +353,142 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+                    Log.LogMessage(MessageImportance.Low, $"Local Symbol path to downloaded file {localSymbolPath}");
+                    List<string> symbolFiles = new List<string>();
+                    symbolFiles.ToList().Add(localSymbolPath);
+                    foreach (var server in serversToPublish)
```
Can you go through this file and add new lines before foreachs, trys, ifs, etc, and after the closing braces.

epananth replied:
I did scan the file a few times and added extra lines..

adiaaida commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/PublishArtifactsInManifest.proj:
```diff
@@ -84,6 +88,7 @@
 
     <Error Condition="'$(ManifestsBasePath)' == ''" Text="ManifestsBasePath is empty. Please provide the full path to asset manifest(s) directory." />
     <Error Condition="'$(BlobBasePath)' == '' OR '$(PackageBasePath)' == ''" Text="A valid full path to BlobBasePath and PackageBasePath is required." />
+    <Error Condition="'$(ArtifactsBasePath)' == ''" Text="ArtifactBasePath is empty. Please provide the full path to ArtifactBasePath directory." />
```
This seems only required if UseStreamingPublishing is true, based on a cursory glance through PublishArtifactsInManifest.

epananth replied:
I removed this and made artifactBasePath to have blobBasePath as a default value, in case the value is empty

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -138,6 +164,16 @@ public abstract class PublishArtifactsInManifestBase : Microsoft.Build.Utilities
         /// </summary>
         public int RetryDelayMilliseconds { get; set; } = 5000;
 
+        public ExponentialRetry RetryHandler = new ExponentialRetry
```
Nit: I think this can be made `readonly`

epananth replied:
added readonly to ExponentialRetry

adiaaida commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestV3.cs:
```diff
@@ -199,16 +189,31 @@ public override async Task<bool> ExecuteAsync()
+                string temporarySymbolsLocation = "";
+                if (!UseStreamingPublishing)
+                {
+
```
Nit: remove new line

epananth replied:
removed new line

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -838,20 +1257,86 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+            Dictionary<string, HashSet<Asset>> buildAssets,
+            TargetFeedConfig feedConfig)
+        {
+            string containerId = await GetContainerIdAsync();
```
If the container ID wasn't found, then an empty string is returned (and an error is logged). We should probably do this after the call, so that the future containerId uses in various places don't do wacky, unexpected things.

```
if (Log.HasLoggedErrors)
{
    return;
}
```

epananth replied:
added has loggedError after the call getContainerId

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -880,15 +1365,160 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
             await blobFeedAction.PushToFeedAsync(packages, pushOptions);
         }
 
+        private async Task PublishPackagesToAzureStorageUsingStreamingAsync(
```
@epananth I think you might be able to make this NYI and eliminate this code altogether. This is for the old blob feed pushes, which I believe we no longer use (I think https://github.com/dotnet/arcade/pull/7086/files#diff-716e74c0c24a454d2c2b3d08566c5c79d62a037bee929a70e53c3f73b62cae2aR657-R672 is dead code now)

Either that, or make a TODO in this method saying it should be removed

epananth replied:
I think even PublishPackagesToAzureStorageNugetByDownloadingEntireFolderAsync is also dead code. I think I can remove that too. So FeedType.AzureStorageFeed is no longer valid for blob upload. 

So if we have FeedType.AzureStorageFeed can I just log warning ?

mmitche replied:
> So if we have FeedType.AzureStorageFeed can I just log warning ?

Yep, I like it.

epananth replied:
removed this code and logged warning

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -880,15 +1365,160 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
+                string temporaryBlobDirectory = CreateTemporaryDirectory();
+                var fileName = Path.GetFileName(blob.Id);
+                var localBlobPath = Path.Combine(temporaryBlobDirectory, fileName);
+                await DownloadFileAsync(client, "BlobArtifacts", containerId, fileName,
```
Removed it and replaced it with an enum

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -331,14 +502,44 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
         {
-            StringBuilder symbolLog = new StringBuilder();
-            symbolLog.AppendLine("Publishing Symbols to Symbol server: ");
+            if (!UseStreamingPublishing)
```
changed it to use 
if(UseStreamingPublishing) 
{}
else
{}

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -103,6 +107,26 @@ public abstract class PublishArtifactsInManifestBase : Microsoft.Build.Utilities
+
+        public bool UseStreamingPublishing { get; set; }
+
+        public int Retries = 5;
```
removed it

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -479,6 +694,136 @@ private HashSet<PackageArtifactModel> FilterPackages(HashSet<PackageArtifactMode
+        /// Creates Azdo client
+        /// </summary>
+        /// <param name="accountName">Name of the org</param>
+        /// <param name="download">If client is used for downloading the artifact then AutomaticDecompression has to be set, else it is not required</param>
```
updated the variable name

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -331,14 +502,44 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+        /// <param name="symbolPublishingExclusionsFile">Right now we do not add any files to this, so this is going to be null</param>
+        /// <param name="temporarySymbolsLocation">Path to Symbol.nupkgs</param>
+        /// <param name="publishSpecialClrFiles">If true, the special coreclr module indexed files like DBI, DAC and SOS are published</param>
+        /// <returns></returns>
```
removed it

epananth commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -317,7 +353,142 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
+
+            if (filesToSymbolServer != null && filesToSymbolServer.Any())
+            {
+                List<string> symbolFiles = null;
```
changed it

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -349,10 +572,13 @@ public void CheckForStableAssetsInNonIsolatedFeeds()
                 {
-                    var pdbEntries = System.IO.Directory.EnumerateFiles(pdbArtifactsBasePath, "*.pdb", System.IO.SearchOption.AllDirectories);
-                    var dllEntries = System.IO.Directory.EnumerateFiles(pdbArtifactsBasePath, "*.dll", System.IO.SearchOption.AllDirectories);
+                    var pdbEntries = System.IO.Directory.EnumerateFiles(pdbArtifactsBasePath, "*.pdb",
```
[nit] More chopping of the parameters at a random place, let's try to stick to one of two ways, which seems to be what the rest of the codebase uses:

* If line not too long, all parameters in same line
* Else one parameter per line

riarenas replied:
Oh, it actually even looks like this is the last remaining instance of this. Nice! this one looks like it was just missed.

epananth replied:
I made this change. I somehow missed this one earlier.

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -974,6 +1666,111 @@ private BlobFeedAction CreateBlobFeedAction(TargetFeedConfig feedConfig)
+        /// Create Temporary directory if it does not exists.
+        /// </summary>
+        /// <param name="temporaryLocation"></param>
+        public void EnsureTemporaryDirectoryExists(string temporaryLocation)
```
I wonder if all these delete methods really need to be public? they seem to be dealing with cleaning up internal implementations of the publishing approach, so they make more sense as private helpers?

epananth replied:
The existing tests will not recognize it if I make them private, so I have them as public.

riarenas reviewed (approved):
I only have a couple more nits and comments but provided you have some test builds with these latest changes that we can take a look at, this LGTM,

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -774,31 +818,38 @@ private async Task<string> GetContainerIdAsync()
         /// </summary>
         /// <param name="client">Azdo client</param>
-        /// <param name="artifact">If it is PackageArtifacts or BlobArtifacts</param>
+        /// <param name="ArtifactName">If it is PackageArtifacts or BlobArtifacts</param>
```
Nit: the parameter name is `artifactName` (casing)

epananth replied:
Updated this

mmitche commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs:
```diff
@@ -1326,121 +1472,26 @@ public void SplitArtifactsInCategories(BuildModel buildModel)
-
-        }
-
-        private async Task PublishPackagesToAzureStorageNugetByDownloadingEntireFolderAsync(HashSet<PackageArtifactModel> packagesToPublish,
```
YAY!

mmitche reviewed (approved):
LGTM. Thanks for iterating on this. Complicated stuff.

adiaaida reviewed (approved):
I'm happy with how it looks in the staging pipeline (the only failures make sense/are because the files had already been uploaded since we tried on a bar id that already has been published). LGTM

epananth commented:
Latest test against runtime 

| Publishing | Link | Time taken to publish |
|-------------|------|-------------------------|
| New infra Bar Id - 85015       |https://dev.azure.com/dnceng/internal/_build/results?buildId=1073502&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a | 24mins 56s | 
| old infra   Bar Id - 85015           | https://dev.azure.com/dnceng/internal/_build/results?buildId=1072291&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a | 23mins 44secs + 3 mins 48 secs |
| New infra (bar id - 84977) | https://dev.azure.com/dnceng/internal/_build/results?buildId=1073701&view=logs&j=33961d73-1e6b-5306-1d3b-c6c083bff7de&t=577fb5b9-bfb7-5430-eb7a-5c5120b89b4d |  25 mins 51 sec| 
| old infra (bar id - 84977) | https://dev.azure.com/dnceng/internal/_build/results?buildId=1071605&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a| 31 mins 32 sec + 3 min 8 sec |

epananth commented:
I did some test against installer and I am currently facing some IO exception, used by another process for blob publishing. I am investigating this issue, so I am not going to merge this PR till that issue is fixed.

epananth commented:
Aspnet core -> https://dev.azure.com/dnceng/internal/_build/results?buildId=1079097&view=results
Installer -> https://dev.azure.com/dnceng/internal/_build/results?buildId=1079093&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a
Runtime -> https://dev.azure.com/dnceng/internal/_build/results?buildId=1079083&view=logs&s=380ab8c6-4139-55ff-15c4-fb48260f4dca&j=ba23343f-f710-5af9-782d-5bd26b102304

epananth commented:
In the last change, I added the following 
1) Added await to CompareLocalPackageToFeedPackage
2) Making sure PackageArchiveReader was using the correct scope
3) added retry to ContainerIdAsync
4) More logs for downloading file 
5) Updated the test

epananth commented:
@mmitche mentioned my change is good to merge. :)

epananth commented:
Installer build first time - https://dev.azure.com/dnceng/internal/_build/results?buildId=1079093&view=logs&j=ba23343f-f710-5af9-782d-5bd26b102304&t=6e277ba4-1c1e-552d-b96f-db0aeb4be20a

did delete-build-from-channel and retried build - https://dev.azure.com/dnceng/internal/_build/results?buildId=1078127&view=results
